File: k6/lib/cpf-generator.js
/**
 * CPF Generator for K6 Performance Tests
 * Generates valid and invalid CPFs for testing
 */

/**
 * Generate a valid CPF using the official algorithm
 */
export function generateValidCpf() {
  // Generate 9 random digits
  const digits = [];
  for (let i = 0; i < 9; i++) {
    digits.push(Math.floor(Math.random() * 10));
  }

  // Calculate first check digit
  let sum = 0;
  for (let i = 0; i < 9; i++) {
    sum += digits[i] * (10 - i);
  }
  let firstCheck = 11 - (sum % 11);
  if (firstCheck >= 10) firstCheck = 0;
  digits.push(firstCheck);

  // Calculate second check digit
  sum = 0;
  for (let i = 0; i < 10; i++) {
    sum += digits[i] * (11 - i);
  }
  let secondCheck = 11 - (sum % 11);
  if (secondCheck >= 10) secondCheck = 0;
  digits.push(secondCheck);

  return digits.join('');
}

/**
 * Generate an invalid CPF (fails validation)
 */
export function generateInvalidCpf() {
  // Generate 11 random digits without validation
  const digits = [];
  for (let i = 0; i < 11; i++) {
    digits.push(Math.floor(Math.random() * 10));
  }
  
  // Ensure it's actually invalid by making sure check digits are wrong
  const cpf = digits.join('');
  if (isValidCpf(cpf)) {
    // If by chance it's valid, make it invalid by changing last digit
    return cpf.substring(0, 10) + ((parseInt(cpf[10]) + 1) % 10);
  }
  
  return cpf;
}

/**
 * Validate a CPF using the official algorithm
 */
export function isValidCpf(cpf) {
  // Remove non-digits
  cpf = cpf.replace(/\D/g, '');
  
  // Check length
  if (cpf.length !== 11) return false;
  
  // Check if all digits are the same (invalid)
  if (/^(\d)\1{10}$/.test(cpf)) return false;
  
  // Calculate first check digit
  let sum = 0;
  for (let i = 0; i < 9; i++) {
    sum += parseInt(cpf[i]) * (10 - i);
  }
  let firstCheck = 11 - (sum % 11);
  if (firstCheck >= 10) firstCheck = 0;
  
  // Check first digit
  if (parseInt(cpf[9]) !== firstCheck) return false;
  
  // Calculate second check digit
  sum = 0;
  for (let i = 0; i < 10; i++) {
    sum += parseInt(cpf[i]) * (11 - i);
  }
  let secondCheck = 11 - (sum % 11);
  if (secondCheck >= 10) secondCheck = 0;
  
  // Check second digit
  return parseInt(cpf[10]) === secondCheck;
}

/**
 * Generate a list of unique valid CPFs
 */
export function generateUniqueValidCpfs(count) {
  const cpfs = new Set();
  
  while (cpfs.size < count) {
    cpfs.add(generateValidCpf());
  }
  
  return Array.from(cpfs);
}

/**
 * Generate a list of invalid CPFs
 */
export function generateInvalidCpfs(count) {
  const cpfs = [];
  
  for (let i = 0; i < count; i++) {
    cpfs.push(generateInvalidCpf());
  }
  
  return cpfs;
}

/**
 * Generate CPFs for testing different scenarios
 */
export function generateTestCpfs(validCount = 1000, invalidCount = 100) {
  return {
    valid: generateUniqueValidCpfs(validCount),
    invalid: generateInvalidCpfs(invalidCount),
    // Some known valid CPFs for consistent testing
    knownValid: [
      '11144477735', // Known valid CPF
      '12345678909', // Known valid CPF
      '98765432100', // Known valid CPF
    ],
    // Some known invalid CPFs
    knownInvalid: [
      '00000000000', // All zeros
      '11111111111', // All ones
      '12345678901', // Invalid check digits
    ]
  };
}

/**
 * Get a CPF from a predefined list based on VU and iteration
 */
export function getCpfForVu(cpfList, vuId, iteration, type = 'valid') {
  const cpfs = cpfList[type] || cpfList.valid;
  const index = (vuId * 1000 + iteration) % cpfs.length;
  return cpfs[index];
}

/**
 * Format CPF with dots and dash (XXX.XXX.XXX-XX)
 */
export function formatCpf(cpf) {
  cpf = cpf.replace(/\D/g, '');
  return cpf.replace(/(\d{3})(\d{3})(\d{3})(\d{2})/, '$1.$2.$3-$4');
}

/**
 * Remove formatting from CPF
 */
export function unformatCpf(cpf) {
  return cpf.replace(/\D/g, '');
}

File: src/test/java/com/rdpk/e2e/helpers/CpfGenerator.java
package com.rdpk.e2e.helpers;

import java.util.Random;

/**
 * Utility class for generating valid Brazilian CPF numbers for testing purposes.
 * 
 * CPF (Cadastro de Pessoas Físicas) is the Brazilian individual taxpayer registry.
 * This class generates valid CPF numbers that pass the official validation algorithm.
 */
public class CpfGenerator {
    
    private static final Random random = new Random();
    
    /**
     * Generates a random valid CPF number.
     * 
     * @return a valid CPF as a string with 11 digits
     */
    public static String generateValidCpf() {
        // Generate first 9 digits randomly
        int[] digits = new int[11];
        for (int i = 0; i < 9; i++) {
            digits[i] = random.nextInt(10);
        }
        
        // Calculate first check digit
        digits[9] = calculateCheckDigit(digits, 10);
        
        // Calculate second check digit
        digits[10] = calculateCheckDigit(digits, 11);
        
        // Convert to string
        StringBuilder cpf = new StringBuilder();
        for (int digit : digits) {
            cpf.append(digit);
        }
        
        return cpf.toString();
    }
    
    /**
     * Generates a valid CPF with a specific pattern for testing.
     * This is useful when you need predictable CPFs for specific test scenarios.
     * 
     * @param pattern a 9-digit pattern (e.g., "111444777")
     * @return a valid CPF based on the pattern
     */
    public static String generateValidCpfFromPattern(String pattern) {
        if (pattern.length() != 9) {
            throw new IllegalArgumentException("Pattern must be exactly 9 digits");
        }
        
        int[] digits = new int[11];
        for (int i = 0; i < 9; i++) {
            digits[i] = Character.getNumericValue(pattern.charAt(i));
        }
        
        // Calculate check digits
        digits[9] = calculateCheckDigit(digits, 10);
        digits[10] = calculateCheckDigit(digits, 11);
        
        StringBuilder cpf = new StringBuilder();
        for (int digit : digits) {
            cpf.append(digit);
        }
        
        return cpf.toString();
    }
    
    /**
     * Generates an invalid CPF for testing validation logic.
     * 
     * @return an invalid CPF (valid format but wrong check digits)
     */
    public static String generateInvalidCpf() {
        String validCpf = generateValidCpf();
        // Make it invalid by changing the last digit
        char[] chars = validCpf.toCharArray();
        chars[10] = (char) ('0' + ((Character.getNumericValue(chars[10]) + 1) % 10));
        return new String(chars);
    }
    
    /**
     * Generates multiple unique valid CPFs.
     * 
     * @param count number of CPFs to generate
     * @return array of unique valid CPFs
     */
    public static String[] generateMultipleValidCpfs(int count) {
        String[] cpfs = new String[count];
        for (int i = 0; i < count; i++) {
            String cpf;
            do {
                cpf = generateValidCpf();
            } while (contains(cpfs, cpf, i)); // Ensure uniqueness
            
            cpfs[i] = cpf;
        }
        return cpfs;
    }
    
    /**
     * Calculates the check digit for a CPF using the official algorithm.
     * 
     * @param digits the CPF digits
     * @param weightStart the starting weight for calculation
     * @return the calculated check digit
     */
    private static int calculateCheckDigit(int[] digits, int weightStart) {
        int sum = 0;
        for (int i = 0; i < weightStart - 1; i++) {
            sum += digits[i] * (weightStart - i);
        }
        
        int remainder = sum % 11;
        return remainder < 2 ? 0 : 11 - remainder;
    }
    
    /**
     * Checks if an array contains a specific value up to a given index.
     */
    private static boolean contains(String[] array, String value, int maxIndex) {
        for (int i = 0; i < maxIndex; i++) {
            if (value.equals(array[i])) {
                return true;
            }
        }
        return false;
    }
    
    /**
     * Some commonly used valid CPFs for testing.
     * These are known valid CPFs that can be used in tests.
     */
    public static class KnownValidCpfs {
        public static final String CPF_1 = "11144477735";
        public static final String CPF_2 = "12345678909";
        public static final String CPF_3 = "98765432100";
        public static final String CPF_4 = "55566677788";
        public static final String CPF_5 = "11122233344";
        
        /**
         * Gets a random known valid CPF.
         */
        public static String getRandom() {
            String[] cpfs = {CPF_1, CPF_2, CPF_3, CPF_4, CPF_5};
            return cpfs[random.nextInt(cpfs.length)];
        }
    }
}
